### PR TITLE
feat(fip0081): initialPledge calculation & ramp, plus migration tests

### PR DESCRIPTION
**Pull request by rvagg**
Initial draft, need to add some tests.

Depends on:

* https://github.com/filecoin-project/builtin-actors/pull/1580 (which is really just https://github.com/filecoin-project/builtin-actors/pull/1557)
* https://github.com/filecoin-project/go-state-types/pull/314

Also has https://github.com/filecoin-project/lotus/pull/12525 because it was in my path to this.

### CODE REVIEW

**Comment by rvagg** [comment]
@ZenGround0 I'm having a bit of trouble testing pledge in here. I can observe the change in pledge value if I log values of internal calculation during the ramp period: https://github.com/filecoin-project/lotus/blob/72cdd3f19f6fb538f9d41f2e435d6cac84dc7c9e/itests/migration_test.go#L974-L986 (logging not included in this PR, but if I do it in go-state-types where the calculation is done I can see it).

But the value I get is capped by the `InitialPledgeMaxPerByte` value for the power being pledged: https://github.com/filecoin-project/go-state-types/blob/f1e9746dbeaaa0411bf8d704420a584a21656f23/builtin/v15/miner/monies.go#L131-L134

Consistently the cap ends up being ~66% of the uncapped pledge.

Got any suggestions for reducing this? I'm guessing it's to do with baseline power, but I'm actually not sure. How can we budge the state-sourced pledge inputs in an itest?

**Comment by rvagg** [comment]
btw `InitialPledgeMaxPerByte` is a `var` so I can mess with it directly, but that's not exactly ideal for testing, I'd like to be modifying as few runtime pseudo-constants as possible.

**Comment by rvagg** [comment]
I've pushed another commit here that applies the `InitialPledgeMaxPerByte` hack to make it work, then observe the pledge calculation before, during and after activation & ramp period. We also implement a form of `StateMinerInitialPledgeForSector` that pulls in the pledge calculation from miner14, so we can run the calculation with the same inputs and get both pre-FIP0081 and post-FIP0081 values and compare them.

* Before migration: we should see no difference between the two
* After migration and during ramp: we should see an increasing difference between the two calculations
* After the ramp: we should still see an increasing distance because the new calculation includes the 30% simple component

Additionally, by capturing rate of change of this delta during and after the ramp we can look at the average rate of change and assert that it should be less after the ramp. I'd like to apply some kind of % to this but that's probably a bridge too far here. We end up with numbers like `0.000080` during and `0.000030` after the ramp; which seems reasonable to me I think.

In our test, the migration is at epoch 100, and the ramp is for 200 epochs.

<details>
<summary>
Eyeballing the outputs seems reasonable I think: (click to see full log)
</summary>

```
Pledge @7: 112287402045 (pre-fip-0081: 112287402045, rate of change: 0.000000)
Pledge @12: 111917456893 (pre-fip-0081: 111917456893, rate of change: 0.000000)
Pledge @17: 111557577612 (pre-fip-0081: 111557577612, rate of change: 0.000000)
Pledge @22: 111207342519 (pre-fip-0081: 111207342519, rate of change: 0.000000)
Pledge @27: 110866356178 (pre-fip-0081: 110866356178, rate of change: 0.000000)
Pledge @32: 110534261666 (pre-fip-0081: 110534261666, rate of change: 0.000000)
Pledge @37: 110210705121 (pre-fip-0081: 110210705121, rate of change: 0.000000)
Pledge @42: 109895360296 (pre-fip-0081: 109895360296, rate of change: 0.000000)
Pledge @47: 109587922739 (pre-fip-0081: 109587922739, rate of change: 0.000000)
Pledge @52: 109288100733 (pre-fip-0081: 109288100733, rate of change: 0.000000)
Pledge @57: 108995627567 (pre-fip-0081: 108995627567, rate of change: 0.000000)
Pledge @62: 108710238299 (pre-fip-0081: 108710238299, rate of change: 0.000000)
Pledge @67: 108431694581 (pre-fip-0081: 108431694581, rate of change: 0.000000)
Pledge @72: 108159760324 (pre-fip-0081: 108159760324, rate of change: 0.000000)
Pledge @77: 107894220765 (pre-fip-0081: 107894220765, rate of change: 0.000000)
Pledge @82: 107634860856 (pre-fip-0081: 107634860856, rate of change: 0.000000)
Pledge @87: 107381484875 (pre-fip-0081: 107381484875, rate of change: 0.000000)
Pledge @92: 107133910661 (pre-fip-0081: 107133910661, rate of change: 0.000000)
Pledge @97: 106891951580 (pre-fip-0081: 106891951580, rate of change: 0.000000)
Pledge @102: 106657061673 (pre-fip-0081: 106655441404, rate of change: 0.000015)
Pledge @107: 106429859252 (pre-fip-0081: 106424219774, rate of change: 0.000038)
Pledge @112: 106208620960 (pre-fip-0081: 106198127046, rate of change: 0.000046)
Pledge @117: 105991977308 (pre-fip-0081: 105977015703, rate of change: 0.000042)
Pledge @123: 105781820052 (pre-fip-0081: 105760748250, rate of change: 0.000058)
Pledge @127: 105574809671 (pre-fip-0081: 105549187056, rate of change: 0.000044)
Pledge @132: 105373440009 (pre-fip-0081: 105342198665, rate of change: 0.000054)
Pledge @137: 105176456011 (pre-fip-0081: 105139667030, rate of change: 0.000053)
Pledge @142: 104984473239 (pre-fip-0081: 104941468601, rate of change: 0.000060)
Pledge @147: 104796079148 (pre-fip-0081: 104747490696, rate of change: 0.000054)
Pledge @152: 104613022985 (pre-fip-0081: 104557627413, rate of change: 0.000066)
Pledge @157: 104432749958 (pre-fip-0081: 104371768356, rate of change: 0.000054)
Pledge @162: 104257812778 (pre-fip-0081: 104189820021, rate of change: 0.000068)
Pledge @167: 104086321831 (pre-fip-0081: 104011683952, rate of change: 0.000065)
Pledge @172: 103920091497 (pre-fip-0081: 103837269258, rate of change: 0.000080)
Pledge @177: 103755938794 (pre-fip-0081: 103666480122, rate of change: 0.000065)
Pledge @182: 103596694433 (pre-fip-0081: 103499236824, rate of change: 0.000079)
Pledge @187: 103440187312 (pre-fip-0081: 103335454796, rate of change: 0.000072)
Pledge @192: 103288036699 (pre-fip-0081: 103175055058, rate of change: 0.000082)
Pledge @197: 103138411264 (pre-fip-0081: 103017960688, rate of change: 0.000074)
Pledge @202: 102992999292 (pre-fip-0081: 102864098367, rate of change: 0.000084)
Pledge @207: 102850876753 (pre-fip-0081: 102713399532, rate of change: 0.000085)
Pledge @212: 102712054901 (pre-fip-0081: 102565790484, rate of change: 0.000088)
Pledge @217: 102575529912 (pre-fip-0081: 102421207554, rate of change: 0.000081)
Pledge @222: 102444360649 (pre-fip-0081: 102279590618, rate of change: 0.000104)
Pledge @227: 102316042044 (pre-fip-0081: 102140875798, rate of change: 0.000104)
Pledge @232: 102189148431 (pre-fip-0081: 102004994925, rate of change: 0.000090)
Pledge @237: 102065207769 (pre-fip-0081: 101871899637, rate of change: 0.000092)
Pledge @243: 101945359778 (pre-fip-0081: 101741529895, rate of change: 0.000106)
Pledge @247: 101825268820 (pre-fip-0081: 101613832989, rate of change: 0.000077)
Pledge @252: 101711260765 (pre-fip-0081: 101488760499, rate of change: 0.000112)
Pledge @257: 101597547746 (pre-fip-0081: 101366255061, rate of change: 0.000089)
Pledge @262: 101488199043 (pre-fip-0081: 101246272784, rate of change: 0.000108)
Pledge @267: 101379885717 (pre-fip-0081: 101128763250, rate of change: 0.000094)
Pledge @272: 101275113248 (pre-fip-0081: 101013681743, rate of change: 0.000105)
Pledge @277: 101172622745 (pre-fip-0081: 100900985386, rate of change: 0.000104)
Pledge @282: 101073224621 (pre-fip-0081: 100790628404, rate of change: 0.000112)
Pledge @287: 100976465944 (pre-fip-0081: 100682572260, rate of change: 0.000115)
Pledge @292: 100881899300 (pre-fip-0081: 100576771176, rate of change: 0.000115)
Pledge @297: 100788825875 (pre-fip-0081: 100473188634, rate of change: 0.000108)
Pledge @302: 100695446438 (pre-fip-0081: 100371785022, rate of change: 0.000083)
Pledge @307: 100598616335 (pre-fip-0081: 100272523531, rate of change: 0.000027)
Pledge @312: 100503654685 (pre-fip-0081: 100175367994, rate of change: 0.000025)
Pledge @317: 100411389989 (pre-fip-0081: 100080284966, rate of change: 0.000031)
Pledge @322: 100320283567 (pre-fip-0081: 99987237319, rate of change: 0.000022)
Pledge @327: 100231590408 (pre-fip-0081: 99896194392, rate of change: 0.000027)
Pledge @332: 100145690433 (pre-fip-0081: 99807124744, rate of change: 0.000035)
Pledge @337: 100061283251 (pre-fip-0081: 99719994422, rate of change: 0.000030)
Pledge @342: 99979384487 (pre-fip-0081: 99634774988, rate of change: 0.000036)
Pledge @347: 99898922730 (pre-fip-0081: 99551434585, rate of change: 0.000032)
Pledge @352: 99819877961 (pre-fip-0081: 99469944272, rate of change: 0.000027)
Pledge @357: 99742027202 (pre-fip-0081: 99390275441, rate of change: 0.000021)
Pledge @362: 99667578344 (pre-fip-0081: 99312405318, rate of change: 0.000037)
Pledge @367: 99595470498 (pre-fip-0081: 99236305052, rate of change: 0.000043)
Pledge @372: 99524070493 (pre-fip-0081: 99161945210, rate of change: 0.000033)
Pledge @377: 99456943315 (pre-fip-0081: 99089308050, rate of change: 0.000058)
Pledge @382: 99389482781 (pre-fip-0081: 99018359088, rate of change: 0.000038)
Pledge @387: 99322093530 (pre-fip-0081: 98949075062, rate of change: 0.000022)
Pledge @392: 99257313876 (pre-fip-0081: 98881437976, rate of change: 0.000031)
Pledge @397: 99192198136 (pre-fip-0081: 98815418544, rate of change: 0.000012)
Pledge @402: 99131380972 (pre-fip-0081: 98751004580, rate of change: 0.000039)
Pledge @407: 99070006323 (pre-fip-0081: 98688163669, rate of change: 0.000017)
Pledge @412: 99012279230 (pre-fip-0081: 98626883848, rate of change: 0.000038)
Pledge @417: 98954166803 (pre-fip-0081: 98567135438, rate of change: 0.000019)
Pledge @422: 98900201346 (pre-fip-0081: 98508908331, rate of change: 0.000046)
Pledge @427: 98844325497 (pre-fip-0081: 98452169159, rate of change: 0.000011)
Pledge @432: 98791419784 (pre-fip-0081: 98396909612, rate of change: 0.000026)
Pledge @437: 98740885551 (pre-fip-0081: 98343109530, rate of change: 0.000035)
Pledge @442: 98691399539 (pre-fip-0081: 98290747504, rate of change: 0.000031)
Pledge @447: 98643316900 (pre-fip-0081: 98239806357, rate of change: 0.000031)
Pledge @452: 98596985476 (pre-fip-0081: 98190269328, rate of change: 0.000035)
Pledge @457: 98551112122 (pre-fip-0081: 98142116240, rate of change: 0.000025)
Pledge @462: 98506778075 (pre-fip-0081: 98095332815, rate of change: 0.000027)
Pledge @467: 98462706312 (pre-fip-0081: 98049899643, rate of change: 0.000016)
Pledge @472: 98422474245 (pre-fip-0081: 98005809126, rate of change: 0.000041)
Pledge @477: 98382647884 (pre-fip-0081: 97963037484, rate of change: 0.000032)
Pledge @482: 98344111036 (pre-fip-0081: 97921571462, rate of change: 0.000032)
Pledge @487: 98305438712 (pre-fip-0081: 97881392624, rate of change: 0.000017)
Pledge @492: 98269268574 (pre-fip-0081: 97842492620, rate of change: 0.000030)
Pledge @497: 98235045247 (pre-fip-0081: 97804855780, rate of change: 0.000037)
Pledge @502: 98200137592 (pre-fip-0081: 97768461654, rate of change: 0.000017)

Average rate of change during ramp: 0.000080
Average rate of change after ramp: 0.000027
```
</details>

Plotting the whole series looks like this:

Screenshot 2024-09-30 at 9 51 53 PM

Also the "rate of change", which is more like "rate of change in the delta between pre and post FIP-0081 calculation", we can see the ramp at work:

Screenshot 2024-09-30 at 9 49 28 PM

**Comment by ZenGround0** [comment]
Thanks for the nice analysis I'll take a look at the code.  

If you wanted to think more about avoiding the cap resetting thing the levers you can play with are 
* making sector size smaller
* making network power bigger 
* moving forward in time so network baseline is bigger
* making the circulating supply smaller 
But your approach seems ok as it is.

File: chain/consensus/filcns/upgrades.go
Comment (ZenGround0) [inline chain/consensus/filcns/upgrades.go:2609]: I think I understand why we are doing a pre upgrade but lets make sure we're thinking the same thing.

The only reason is to speed up the base migration since the power actor state is negligible.

The premigration speeds up the migration by writing all 3M actors with new code cids to badger and only a small fraction of those will change.  So our badger writes will speed up.  But we don't actually cache base migration changes in memory so we're not gaining anything from caching.

@rvagg does this line up with your thinking?

File: chain/consensus/filcns/upgrades.go
Comment (ZenGround0) [inline chain/consensus/filcns/upgrades.go:2620]: Since you're adding this to the general migration config please leave an issue in go-state-types for removing these fields after nv24.

File: chain/consensus/filcns/upgrades.go
Comment (ZenGround0) [inline chain/consensus/filcns/upgrades.go:2692]: Let's hope no one ever sees it but please fix this error

File: chain/gen/genesis/miners.go
Comment (ZenGround0) [inline chain/gen/genesis/miners.go:506]: Can't you just set this to 0,0 and avoid adding all of this complexity?  This would use the same argument you made in PR comments that differences in pledge are small and there is plenty of buffer.

Comment (ZenGround0) [reply chain/gen/genesis/miners.go:506]: 0 epochs since start, 0 ramp duration should put us at the end of the ramp

Comment (ZenGround0) [reply chain/gen/genesis/miners.go:506]: ah you know what, if you don't change it to this then its a bug.  The preseal setup actors stub we left for genesis generation which gets called below assumes ramp time == 0. https://github.com/filecoin-project/builtin-actors/pull/1557/files#diff-fbcb2ec1a9d82b18f146c728cafd643df0e7ae47a04d84be7644913fe89236e5R1988

File: go.mod
Comment (ZenGround0) [inline go.mod:52]: Reminder to update to the tagged value when its out

File: node/impl/full/state.go
Comment (ZenGround0) [inline node/impl/full/state.go:1571]: I'll advise again that it would be simpler to just have a function `max(InitialPledgeForPower(..., -1,0,...), InitialPledgeForPower(..., 0,0,...)` that takes max before and after ramp that you can use for these Miner APIs.

But what you have looks fine too.

Comment (ZenGround0) [reply node/impl/full/state.go:1571]: And I guess this is better for testing

File: itests/kit/node_unmanaged.go
Comment (ZenGround0) [inline itests/kit/node_unmanaged.go:791]: 🎉

Comment (ZenGround0) [reply itests/kit/node_unmanaged.go:791]: My favorite file in this PR

File: itests/migration_test.go
Comment (ZenGround0) [inline itests/migration_test.go:978]: Its worth documenting somewhere that this test relies on network power being below baseline.  The impact is in the opposite direction if we are well above baseline

File: itests/migration_test.go
Comment (ZenGround0) [inline itests/migration_test.go:991]: can probably remove now that is working

**Review by ZenGround0 — COMMENTED** [review]
Overall LGTM I'll do a final review once all dependencies are landed before approving

File: itests/migration_test.go
Comment (ZenGround0) [reply itests/migration_test.go:978]: My mistake -- not opposite direction but no change if above baseline

File: build/buildconstants/params_calibnet.go
Comment (ZenGround0) [reply build/buildconstants/params_calibnet.go:110]: 3 is also fine

File: chain/consensus/filcns/upgrades.go
Comment (ZenGround0) [reply chain/consensus/filcns/upgrades.go:2620]: code comments will also help

File: chain/consensus/filcns/upgrades.go
Comment (rvagg) [reply chain/consensus/filcns/upgrades.go:2609]: Yes. I decided that it would be best to document this in the code, so I added a comment at the top of this function for future readers.

File: chain/consensus/filcns/upgrades.go
Comment (rvagg) [reply chain/consensus/filcns/upgrades.go:2620]: removed from config struct, added to method params but also added comments

File: build/builtin_actors_gen.go
Comment (rvagg) [reply build/builtin_actors_gen.go:169]: I'll remove the commit that added these before landing this PR; but it'll depend on getting a new bundle in before landing this PR (FYI @rjan90)

File: go.mod
Comment (rvagg) [reply go.mod:52]: I will wait for a tag before I update this PR and merge it. That will depend on getting https://github.com/filecoin-project/go-state-types/pull/314 merged (FYI @rjan90).

File: itests/migration_test.go
Comment (rvagg) [reply itests/migration_test.go:991]: I don't mind leaving sanity-check items like this in place, it serves as good documentation and makes it future-proof too if someone wants to tinker.

File: itests/migration_test.go
Comment (rvagg) [reply itests/migration_test.go:978]: 👌 good point, documented right here

File: itests/migration_test.go
Comment (ZenGround0) [inline itests/migration_test.go:951]: It doesn't look like this PR actually sets this struct for the main lotus node yet.  But when we do shouldn't Expensive be false ?  Looking at how this works maybe all upgrades are safe to consider as expensive?

**Review by ZenGround0 — APPROVED** [review]

File: itests/migration_test.go
Comment (rvagg) [reply itests/migration_test.go:951]: It's already set in https://github.com/filecoin-project/lotus/pull/12455 and I think it's always `true` where there's a state migration involved so we can do things like reject random "execute messages" API requests where they cross the migration boundary. See `git grep -i ErrExpensiveFork` for where this is used.

**Comment by rvagg** [comment]
Removed my commit with the custom actors build and added the tagged build from https://github.com/filecoin-project/builtin-actors/releases/tag/v15.0.0-rc1; also updated go-state-types to `@master`, once we get a tag I'll update that commit here to the tagged version and then this should be good to land.

**Pull request merged**
